staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto

```diff
@@ -244,6 +244,9 @@ message Mount {
     // - when set to true, readonly must be explicitly set to true, and propagation must be PRIVATE (0).
     // - (readonly == false && recursive_read_only == false) does not make the mount read-only.
     bool recursive_read_only = 8;
+    // Mount an image reference (image ID or digest), which is a special use
+    // case for OCI volume mounts.
+    string image_ref = 9;
```

Comment:
at first glance the inclusion of only image_ref seemed lossy (we lost information about the path of the source mount), but I actually think this is a clean way to specify the mount path: ask the CRI to mount the image in from wherever that imageRef is stored. I almost wonder if we need mount_ref at all now...

One thing to note: if we do keep mount_ref, and intend on using it for [host path](https://github.com/kubernetes/kubernetes/pull/125659/files#diff-7a3c2723dedc6b44cf183fc52f84cc4015ffbd65a998a65c3959a953cb63c945R227), it's possible we won't have that for the CoCo case where image_ref is an ID.

also: if mount_ref if specified in the pull image response, does the kubelet always specify image_ref? or do we only do so if mount_ref is empty?

Reply:
The path of the source mount is still available in the host path of the mount. 

I’d like to keep the new mount_ref field, it tells the kubelet that the feature is supported as well as that we have to re-pull on reboot scenarios. 

> if mount_ref is specified in the pull image response, does the kubelet always specify image_ref? or do we only do so if mount_ref is empty?

the image ref is a good hint for telling end users which artifact we pulled. That information should not get lost, so we set both fields (mount_ref and image_ref) for the mount use case. If mount is specified on pull, but no mount_ref got set, then the kubelet falls back to the image_ref in the mount config on container creation. This means that the kubelet cannot know if runtimes support that feature or not, and the mount will be simply not available within the container (seems fine to me).

Reply:
works for me!

Reply:
> This means that the kubelet cannot know if runtimes support that feature or not in that special case and the mount will be simply not available within the container (seems fine to me).

nod this is why I was good with the type field.. to avoid support questions..

Reply:
I don't understand why we removed the mount type field.. It had two uses.. 1: the runtime supports using this image as a mount source for create container mounts.. (should work in image status for ensure, restarts, and ephemeral containers) 2: allows for crio to chose if volume mounting via host path or image ref.  We can probably use the long winded pull image with full response grpc api availability as the indicator for 1, that the container runtime supports the new volume source type being an image.  Problem with reducing to just mount ref as the switch determination, is containerd won't be setting that field in the host cache or cc scenario, whether it has support or not.

Reply:
ok this may be cursed, but what if we had mount_ref == image_ref in the coco/host cache case? as in: it's the field to tell the kubelet how to uniquely refer to this mount to the runtime. Then, we get the compat support (if mount_ref == "", then kubelet knows runtime doesn't support), and we have a way of containerd telling kubelet it should reference the image instead of the mount point. Then, in the host_path field, kubelet would specify the image_ref for both host_path and image_ref (in the coco case).

I think it's the full struct I had a problem with, when we can really represent everything with a single new string

Reply:
@haircommander yes, that’s exactly how it’s implemented in my WIP PR right now:

https://github.com/kubernetes/kubernetes/pull/125663/commits/da75f7c2a5ebb0901f6289cbdbbda0b40f7a8385#diff-e81aa7518bebe9f4412cb375a9008b3481b19ec3e851d3187b3021ee94148f0dR269-R270

https://github.com/kubernetes/kubernetes/pull/125663/commits/da75f7c2a5ebb0901f6289cbdbbda0b40f7a8385#diff-0e1a8d22b0f1dc42d412749c6a916b4c76394ead739d4bfbf26564faedd74a2fR1350

Reply:
ok so mount_ref is a flag indicating support for the new feature in both status and image pull; mount_ref can be host_path or an image_ref based on equivalence check in response from image pull; mount_ref can be image id or image digest[0] based on equivalence checks in response from get image status; image_ref can be id or digest based on container runtime config/convention; volume source for a mount can be the host path field or the image ref field based on image pull result/status and only one should be set per mount entry.

Reply:
oh shoot good point, maybe instead of `image_ref` we should use `image_id`

Reply:
On the struct thing.. I think we're switching from string to pull response struct which was a legacy api definition problem that we didn't use the pull response struct.. And we're adding at least one more field.. My preference is still to define a volume mount type that is supported by the container runtime for this image+runtime+snapshotter fs, as I think that's cleaner than having a field that could be 1 or 2 different types of values depending on some alg/logic.. Remembering confusion issues we've had over image ref/id/digest. Is the confusion worth the extra bytes.

Reply:
I readded the `OCIVolumeMountSource` enum to the `PullImageResponse` to indicate that the feature is unsupported by runtimes.

I also split-up the new fields `image_ref` and `mount_ref` within the `Image` message. This way we can cleanly decide which kind of mount we have during `ImageStatus`.